sys/net/gnrc/transport_layer/tcp/gnrc_tcp_eventloop.c

```diff
@@ -0,0 +1,277 @@
+ * @return zero on success
+ * @return negative value on error
+ * @return -EACCES if not able to get write access to packet
+ * @return -ERANGE if segment offset is less than 5
+ * @return -ENOMSG if paket can't be marked
+ * @return -EINVAL if checksum was invalid
+ * @return -ENOTCONN if no module is interested in this context
+ */
+static int _receive(gnrc_pktsnip_t *pkt)
+{
+    /* NOTE: Receiving direction: pkt = payload, payload->next = tcp, tcp->next = nw */
+    uint16_t ctl = 0;
+    uint16_t src = 0;
+    uint16_t dst = 0;
+    uint8_t hdr_size = 0;
+    uint8_t syn = 0;
+    gnrc_pktsnip_t *ip = NULL;
+    gnrc_pktsnip_t *reset = NULL;
+    gnrc_tcp_tcb_t *tcb = NULL;
+    tcp_hdr_t *hdr;
+
+    /* Mark TCP header */
+    gnrc_pktsnip_t *tcp = gnrc_pktbuf_start_write(pkt);
+    if (tcp == NULL) {
+        DEBUG("gnrc_tcp_eventloop.c : _receive() : can't write to packet\n");
+        gnrc_pktbuf_release(pkt);
+        return -EACCES;
+    }
+    pkt = tcp;
+
+#ifdef MODULE_GNRC_IPV6
+    /* Get IP Header, discard packet if doesn't contain an ip header */
+    LL_SEARCH_SCALAR(pkt, ip, type, GNRC_NETTYPE_IPV6);
+    if (ip == NULL) {
+        DEBUG("gnrc_tcp_eventloop.c : _receive() : pkt contains no IP Header\n");
+        gnrc_pktbuf_release(pkt);
+        return 0;
+    }
+#endif
+
+    /* Get TCP Header */
+    LL_SEARCH_SCALAR(pkt, tcp, type, GNRC_NETTYPE_TCP);
+    if (tcp == NULL) {
+      DEBUG("gnrc_tcp_eventloop.c : _receive() : pkt contains no TCP Header\n");
+      gnrc_pktbuf_release(pkt);
+      return 0;
+    }
+
+    /* Extract control bits */
+    hdr = (tcp_hdr_t *)tcp->data;
+    ctl = byteorder_ntohs(hdr->off_ctl);
+
+    /* Validate Offset */
+    if (GET_OFFSET(ctl) < OPTION_OFFSET_BASE) {
+        DEBUG("gnrc_tcp_eventloop.c : _receive() : unexpected Offset Value\n");
+        gnrc_pktbuf_release(pkt);
+        return -ERANGE;
+    }
+
+    /* Calculate tcp header size */
+    hdr_size = GET_OFFSET(ctl) * 4;
+
+    /* Mark TCP-Header, if it contains any payload */
+    if ((pkt->type == GNRC_NETTYPE_TCP) && (pkt->size != hdr_size)) {
+        tcp = gnrc_pktbuf_mark(pkt, hdr_size, GNRC_NETTYPE_TCP);
+        if (tcp == NULL) {
+            DEBUG("gnrc_tcp_eventloop.c : _receive() : Header marking failed\n");
+            gnrc_pktbuf_release(pkt);
+            return -ENOMSG;
+        }
+        pkt->type = GNRC_NETTYPE_UNDEF;
+    }
+
+    /* Validate Checksum */
+    if (byteorder_ntohs(hdr->checksum) != _pkt_calc_csum(tcp, ip, pkt)) {
+        DEBUG("gnrc_tcp_eventloop.c : _receive() : Invalid checksum\n");
+        gnrc_pktbuf_release(pkt);
+        return -EINVAL;
+    }
+
+    /* De-multiplex connection */
+    syn = ((ctl & MSK_SYN_ACK) == MSK_SYN);
+    src = byteorder_ntohs(hdr->src_port);
+    dst = byteorder_ntohs(hdr->dst_port);
+
+    /* Find tcb to de-multiplex this packet to */
+    mutex_lock(&_list_gnrc_tcp_tcb_lock);
+    tcb = _list_gnrc_tcp_tcb_head;
+    while (tcb) {
+#ifdef MODULE_GNRC_IPV6
```

Comment:
something is wrong with the TCB matching here. I wrote a small tcp server to listen on a certain port on any address and prints incoming messages. I sniffed with wireshark, and saw that handshake was successful. But when I send data to the server the server replies with a RST package (and win=0). 

I enable debug output and saw that no matching TCB was found here. I then commented this whole section out, effectively having `tcb = _list_gnrc_tcp_tcb_head;`, which is okay because my test server can only handle 1 connection. And that worked! 

In conclusion, the checks below do not find the correct tcb, that is none in my test case 😞 

I'll try to find the error and provide a fix ...